backend/danswer/connectors/teams/connector.py

```diff
@@ -0,0 +1,269 @@
+            replies = message.replies.get_all().execute_query()
+
+            post_message_list: list[ChatMessage] = [message]
+            post_message_list.extend(replies)
+
+            post_message_lists.append(post_message_list)
+
+        return post_message_lists
+
+    def get_channel_object_list_from_team_list(
+        self,
+        team_object_list: list[Team],
+        start: datetime | None = None,
+        end: datetime | None = None,
+    ) -> list[Channel]:
+        filter_str = ""
+        if start is not None and end is not None:
+            filter_str = f"last_modified_datetime ge {start.isoformat()} and last_modified_datetime le {end.isoformat()}"
+
+        channel_list: list[Channel] = []
+        for team_object in team_object_list:
+            query = team_object.channels.get()
+            if filter_str:
+                query = query.filter(filter_str)
+            channel_objects = query.execute_query()
+            channel_list.extend(channel_objects)
+
+        return channel_list
+
+    def get_all_team_objects(self) -> list[Team]:
+        if self.graph_client is None:
+            raise ConnectorMissingCredentialError("Teams")
+
+        team_object_list: list[Team] = []
+
+        teams_object = self.graph_client.teams.get().execute_query()
+
+        if len(self.requested_team_list) > 0:
+            for requested_team in self.requested_team_list:
+                adjusted_request_string = requested_team.replace(" ", "")
+                for team_object in teams_object:
+                    adjusted_team_string = team_object.display_name.replace(" ", "")
+                    if adjusted_team_string == adjusted_request_string:
+                        team_object_list.append(team_object)
+        else:
+            team_object_list.extend(teams_object)
+
+        return team_object_list
+
+    def _fetch_from_teams(
+        self, start: datetime | None = None, end: datetime | None = None
+    ) -> GenerateDocumentsOutput:
+        if self.graph_client is None:
+            raise ConnectorMissingCredentialError("Teams")
+
+        team_object_list = self.get_all_team_objects()
+
+        channel_list = self.get_channel_object_list_from_team_list(
+            team_object_list=team_object_list,
+            start=start,
+            end=end,
+        )
+
+        # goes over channels, converts them into Document objects and then yields them in batches
+        doc_batch: list[Document] = []
+        batch_count = 0
+        for channel_object in channel_list:
+            post_message_lists = self.get_post_message_lists_from_channel(
+                channel_object
+            )
+            for base_message_groups in post_message_lists:
+                doc_batch.append(
+                    self.convert_post_message_list_to_document(
+                        channel_object, base_message_groups
+                    )
+                )
+
+            batch_count += 1
```

Comment:
try not to keep counters yourself, it's better to use the objects directly when it's not very inefficient to do so. ie. use the len function instead of keeping count yourself.